node/node.go

```diff
@@ -0,0 +1,314 @@
+			return ErrDatadirUsed
+		}
+		return err
+	}
+	// Start each of the services
+	started := []string{}
+	for id, service := range services {
+		// Start the next service, stopping all previous upon failure
+		if err := service.Start(running); err != nil {
+			for _, id := range started {
+				services[id].Stop()
+			}
+			running.Stop()
+
+			return err
+		}
+		// Mark the service started for potential cleanup
+		started = append(started, id)
+	}
+	// Finish initializing the startup
+	n.services = services
+	n.server = running
+	n.stop = make(chan struct{})
+
+	return nil
+}
+
+// Stop terminates a running node along with all it's services. In the node was
+// not started, an error is returned.
+func (n *Node) Stop() error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	// Short circuit if the node's not running
+	if n.server == nil {
+		return ErrNodeStopped
+	}
+	// Otherwise terminate all the services and the P2P server too
+	failure := &StopError{
+		Services: make(map[string]error),
+	}
+	for id, service := range n.services {
+		if err := service.Stop(); err != nil {
+			failure.Services[id] = err
+		}
+	}
+	n.server.Stop()
+
+	n.services = nil
+	n.server = nil
+	close(n.stop)
+
+	if len(failure.Services) > 0 {
+		return failure
+	}
+	return nil
+}
+
+// Wait blocks the thread until the node is stopped. If the node is not running
+// at the time of invocation, the method immediately returns.
+func (n *Node) Wait() {
+	n.lock.RLock()
+	if n.server == nil {
+		return
+	}
+	stop := n.stop
+	n.lock.RUnlock()
+
+	<-stop
+}
+
+// Restart terminates a running node and boots up a new one in its place. If the
+// node isn't running, an error is returned.
+func (n *Node) Restart() error {
+	if err := n.Stop(); err != nil {
+		return err
+	}
+	if err := n.Start(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Server retrieves the currently running P2P network layer. This method is meant
+// only to inspect fields of the currently running server, life cycle management
+// should be left to this Node entity.
+func (n *Node) Server() *p2p.Server {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+
+	return n.server
+}
+
+// Service retrieves a currently running service registered under a given id.
+func (n *Node) Service(id string) Service {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+
+	// Short circuit if the node's not running
+	if n.server == nil {
+		return nil
+	}
+	return n.services[id]
+}
+
+// SingletonService retrieves a currently running service using a specific type
+// implementing the Service interface. This is a utility function for scenarios
+// where it is known that only one instance of a given service type is running,
+// allowing to access services without needing to know their specific id with
+// which they were registered. Note, this method uses reflection, so do not run
+// in a tight loop.
+func (n *Node) SingletonService(service interface{}) (string, error) {
```

Comment:
I see that makes sense.

Then we should rethink our strategy here on accepting multiple protocols and therefor perhaps remove the `Service(id)` method instead. In fact regardless of what we do that method ought to be removed because, as you say, you've no idea where it's being referenced from once you change the name of the registered protocol (oh the joy of refactoring javascript events).

I can see multiple services of the same type being an actual thing (e.g. consortium chains). Perhaps we should rethink this all together.